src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOpAnalyzer.php

```diff
@@ -240,19 +251,117 @@ public static function analyze(
                     || $stmt instanceof PhpParser\Node\Expr\BinaryOp\GreaterOrEqual
                     || $stmt instanceof PhpParser\Node\Expr\BinaryOp\Smaller
                     || $stmt instanceof PhpParser\Node\Expr\BinaryOp\SmallerOrEqual)
-                && $statements_analyzer->getCodebase()->config->strict_binary_operands
                 && $stmt_left_type
                 && $stmt_right_type
-                && (($stmt_left_type->isSingle() && $stmt_left_type->hasBool())
-                    || ($stmt_right_type->isSingle() && $stmt_right_type->hasBool()))
             ) {
-                IssueBuffer::maybeAdd(
-                    new InvalidOperand(
-                        'Cannot compare ' . $stmt_left_type->getId() . ' to ' . $stmt_right_type->getId(),
-                        new CodeLocation($statements_analyzer, $stmt),
-                    ),
-                    $statements_analyzer->getSuppressedIssues(),
-                );
+                if ($statements_analyzer->getCodebase()->config->strict_binary_operands
+                    && (($stmt_left_type->isSingle() && $stmt_left_type->hasBool())
+                        || ($stmt_right_type->isSingle() && $stmt_right_type->hasBool()))) {
+                    IssueBuffer::maybeAdd(
+                        new InvalidOperand(
+                            'Cannot compare ' . $stmt_left_type->getId() . ' to ' . $stmt_right_type->getId(),
+                            new CodeLocation($statements_analyzer, $stmt),
+                        ),
+                        $statements_analyzer->getSuppressedIssues(),
+                    );
+                } else {
+                    $left_literal_int_floats = self::getMinMaxLiteralIntFloat($stmt_left_type);
+                    $right_literal_int_floats = self::getMinMaxLiteralIntFloat($stmt_right_type);
+
+                    foreach ($stmt_left_type->getAtomicTypes() as $atomic_type) {
+                        if ($atomic_type instanceof TString
+                            || $atomic_type instanceof TInt
+                            || $atomic_type instanceof TArrayKey
+                            || $atomic_type instanceof TFloat
+                            || $atomic_type instanceof TResource) {
+                            continue;
+                        }
+
+                        if (AtomicTypeComparator::isContainedBy(
+                            $statements_analyzer->getCodebase(),
+                            $atomic_type,
+                            new TNamedObject('DateTimeInterface'),
```

Comment:
[GMP](https://www.php.net/manual/en/class.gmp.php) objects also overload operators

Reply:
It looks like the following classes from php-src have custom comparison behaviour:

```console
$ rg '\.compare = ' | rg -v 'zend_objects_not_comparable'

ext/date/php_date.c:	date_object_handlers_date.compare = date_object_compare_date;
ext/date/php_date.c:	date_object_handlers_immutable.compare = date_object_compare_date;
ext/date/php_date.c:	date_object_handlers_timezone.compare = date_object_compare_timezone;
ext/date/php_date.c:	date_object_handlers_interval.compare = date_interval_compare_objects;
Zend/zend_closures.c:	closure_handlers.compare = zend_closure_compare;
ext/gmp/gmp.c:	gmp_object_handlers.compare = gmp_compare;
ext/intl/breakiterator/breakiterator_class.cpp:	BreakIterator_handlers.compare = BreakIterator_compare_objects;
ext/intl/timezone/timezone_class.cpp:	TimeZone_handlers.compare = TimeZone_compare_objects;
ext/simplexml/simplexml.c:	sxe_object_handlers.compare = sxe_objects_compare;
ext/spl/spl_array.c:	spl_handler_ArrayObject.compare = spl_array_compare_objects;
```

Reply:
> According to the docs [php.net/manual/en/class.dateinterval.php](https://www.php.net/manual/en/class.dateinterval.php) this is incomparable - if I understand this correctly we cannot use <>

Correct, it just emits a warning and fails:
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/date/php_date.c#L4273-L4281

> I guess this applies also to
>> ext/date/php_date.c: date_object_handlers_timezone.compare = date_object_compare_timezone;

No, there are actually several defined ways (depending on the TZ type) to compare `DateTimeZone`:
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/date/php_date.c#L2035-L2048

But not `IntlTimezone` - this one is basically unorderable:
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/intl/timezone/timezone_class.cpp#L281-L287 

Closures are unorderable too:
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/Zend/zend_closures.c#L429-L457

Same for BreakIterator (but looks buggy):
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/intl/breakiterator/breakiterator_class.cpp#L88-L92

Ditto for SimpleXMLObject: 
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/simplexml/simplexml.c#L1233-L1250

SplArray compares as array:
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/spl/spl_array.c#L896-L899

For FFI (not in the list above) PHP compares pointers:
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/ffi/ffi.c#L1771

SplObjectStorage (also not in the list) are sorted as arrays:
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/spl/spl_observer.c#L397

GMP is properly comparable:
https://github.com/php/php-src/blob/f2ec6e480615f4a5633eb07ad8bfb7b2cadcb921/ext/gmp/gmp.c#L434

> Should I disable the error for these like I did for DateTimeInterface?

Since we probably don't want to allow comparing arrays (with `> / <`) this leaves:
* DateTime/DateTimeImmutable
* GMP
* DateTimeZone

Yes, please disable the error for them.